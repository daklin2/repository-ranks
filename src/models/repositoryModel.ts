export enum Languages {
  JAVA_SCRIPT = "javascript",
  TYPE_SCRIPT = "typescript",
  JAVA = "java",
  SCALA = "scala",
  RUBY = "ruby",
  PYTHON= "python",
  ALL = "",
}

export interface RepositoryModel {
  login: string;
  avatarUrl: string;
  stars: number;
  repositoryUrl: string;
}