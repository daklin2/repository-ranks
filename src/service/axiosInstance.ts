import axios from "axios";

const axiosInstance = axios.create({
	baseURL: process.env.REACT_APP_GITHUB_URL,
	headers: {
		"Content-Type": "application/json",
		Accept: "application/vnd.github.v3+json",
	},
	timeout: 4000,
});

export default axiosInstance;
