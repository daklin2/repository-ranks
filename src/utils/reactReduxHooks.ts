import { TypedUseSelectorHook, useDispatch as useDispatchOrg, useSelector as useSelectorOrg } from 'react-redux';
import type { AppState, AppDispatch } from '../store/store';

export const useDispatch = () => useDispatchOrg<AppDispatch>();
export const useSelector: TypedUseSelectorHook<AppState> = useSelectorOrg;