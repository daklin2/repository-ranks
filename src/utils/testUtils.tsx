import {Provider} from "react-redux";
import {Store} from "@reduxjs/toolkit";
import {mount} from "enzyme";

export const mountWrapper = (store: Store, element: JSX.Element) => {
  return mount(
    <Provider store={store}>
      {element}
    </Provider>
  )
}