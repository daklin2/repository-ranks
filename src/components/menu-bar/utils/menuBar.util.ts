import {Languages} from "../../../models/repositoryModel";

export const listOfLanguages = [
  {
    displayName: "JavaScript",
    queryName: Languages.JAVA_SCRIPT,
  },
  {
    displayName: "TypeScript",
    queryName: Languages.TYPE_SCRIPT,
  },
  {
    displayName: "Java",
    queryName: Languages.JAVA,
  },
  {
    displayName: "Python",
    queryName: Languages.PYTHON,
  },
  {
    displayName: "Ruby",
    queryName: Languages.RUBY,
  },
  {
    displayName: "Scala",
    queryName: Languages.SCALA,
  },
  {
    displayName: "All",
    queryName: Languages.ALL,
  }
];