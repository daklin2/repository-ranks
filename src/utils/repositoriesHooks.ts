import {changeSelectedLanguage, getLanguageStats} from "../store/languageMenuSlice";
import {useDispatch} from "./reactReduxHooks";
import {Languages} from "../models/repositoryModel";

export const useChangeSelectedLanguage = () => {
  const dispatch = useDispatch();

  return (language: Languages) => {
    dispatch(getLanguageStats(language))
      .then(() => {
        dispatch(changeSelectedLanguage(language))
      })
  }

}