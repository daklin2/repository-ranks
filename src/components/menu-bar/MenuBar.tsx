import "./MenuBar.style.scss";
import {listOfLanguages} from "./utils/menuBar.util";
import MenuBarLanguageElement from "./view/language-element";
import { useSelector} from "../../utils/reactReduxHooks";
import {useMemo} from "react";

const MenuBar = () => {
  const selectedLanguage = useSelector((state) => state.languageMenu.currentLanguage)

  const listOfLanguage = useMemo(() => {
    return listOfLanguages.map((language, i) => {
      return <MenuBarLanguageElement key={language.queryName + `_${i}`} name={language.displayName} queryName={language.queryName} />
    })
  }, [selectedLanguage])

  return (
    <div className={"menu-bar"}>
      {listOfLanguage}
    </div>
  )
}

export default MenuBar;
