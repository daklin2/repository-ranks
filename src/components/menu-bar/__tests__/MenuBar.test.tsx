import {shallow} from "enzyme";
import toJson from "enzyme-to-json";
import * as ReactReduxHooks from "../../../utils/reactReduxHooks";
import thunk from "redux-thunk";
import configureStore from "redux-mock-store";
import MenuBar from "../MenuBar";
import {Languages} from "../../../models/repositoryModel";
import {mountWrapper} from "../../../utils/testUtils";
import {listOfLanguages} from "../utils/menuBar.util";
import axiosInstance from "../../../service/axiosInstance";

const useDispatchSpy = jest.spyOn(ReactReduxHooks, "useDispatch");
useDispatchSpy.mockImplementation(() => defaultStore.dispatch)

const defaultStoreState = {
  languageMenu: {
    currentLanguage: null,
  }
};
const defaultStore = configureStore([thunk])(defaultStoreState);

describe("MenuBar", () => {
  const useSelectorSpy = jest.spyOn(ReactReduxHooks, "useSelector");
  const useDispatchSpy = jest.spyOn(ReactReduxHooks, "useDispatch");

  beforeEach(() => {
    useSelectorSpy.mockImplementation((state) => state(defaultStore.getState() as any));
    useDispatchSpy.mockImplementation(() => defaultStore.dispatch);
  });
  afterEach(() => {
    jest.clearAllMocks();
    defaultStore.clearActions();
  });

  it("should render MenuBar", () => {
    expect(toJson(shallow(<MenuBar/>))).toMatchSnapshot();
  });

  it("selected language should has active class", () => {
    const editedStoreState = {
      languageMenu: {
        ...defaultStoreState,
        currentLanguage: Languages.TYPE_SCRIPT,
      }
    }
    const store = configureStore([thunk])(editedStoreState);
    useSelectorSpy.mockImplementation((state) => state(store.getState() as any));
    const wrapper = mountWrapper(store, <MenuBar />);

    expect(wrapper.find('[className="menu-bar__language menu-bar__language--active"]').length).toEqual(1)
  });
})