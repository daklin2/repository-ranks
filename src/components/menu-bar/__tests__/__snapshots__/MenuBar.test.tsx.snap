// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuBar should render MenuBar 1`] = `
<div
  className="menu-bar"
>
  <MenuBarLanguageElement
    key="javascript_0"
    name="JavaScript"
    queryName="javascript"
  />
  <MenuBarLanguageElement
    key="typescript_1"
    name="TypeScript"
    queryName="typescript"
  />
  <MenuBarLanguageElement
    key="java_2"
    name="Java"
    queryName="java"
  />
  <MenuBarLanguageElement
    key="python_3"
    name="Python"
    queryName="python"
  />
  <MenuBarLanguageElement
    key="ruby_4"
    name="Ruby"
    queryName="ruby"
  />
  <MenuBarLanguageElement
    key="scala_5"
    name="Scala"
    queryName="scala"
  />
  <MenuBarLanguageElement
    key="_6"
    name="All"
    queryName=""
  />
</div>
`;
